keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,938 @@
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+None.
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+- The feature can be disabled by setting the feature-gate to false if the performance impact of it is not tolerable.
+- The peer-to-peer connection between API servers should be checked to ensure that the remote API servers are reachable from a given API server
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+
+## Drawbacks
+
+<!--
+Why should this KEP _not_ be implemented?
+-->
+
+## Alternatives
+
+### Network location of apiservers
+
+1. Use [endpoint reconciler's masterlease](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/controlplane/reconcilers/lease.go) 
+    1. We will use the already existing IP in Endpoints.Subsets.Addresses of the masterlease by default
+    2. For users with network configurations that would not allow Endpoints.Subsets.Addresses to be reachable from a kube-apiserver, we will introduce a new optional --bind-peer-ip flag to kube-apiserver. We will store its value as an annotation on the masterlease and use this to route the request to the right destination server
+    3. We will also need to store the apiserver identity as an annotation in the masterlease so that we can map the identity of the apiserver to its IP
+    4. We will also expose the IP and port information of the kube-apiservers as annotations in APIserver identity lease object for visibility/debugging purposes
+
+* Pros
+  1. Masterlease reconciler already stores kube-apiserver IPs currently
+  2. This information is not exposed to users in an API that can be used maliciously
+  3. Existing code to handle lifecycle of the masterleases is convenient
+
+* Cons
+  1. using masterlease will include making some changes to the legacy code that does the endpoint reconciliation which is known to be brittle
```

Comment:
Could we use a separate reconciliation loop that is only enabled when our alpha feature gate is turned on?

Reply:
Updated the decision under [Identifying destination apiserver's network location](https://github.com/kubernetes/enhancements/blob/01637e17da9a0d1359a9304ad005c88234cd61ca/keps/sig-api-machinery/4020-unknown-version-interoperability-proxy/README.md#identifying-destination-apiservers-network-location) section.